sklearn/utils/estimator_checks.py

```diff
@@ -4012,3 +4013,52 @@ def check_transformer_get_feature_names_out_pandas(name, transformer_orig):
     assert (
         len(feature_names_out_default) == n_features_out
     ), f"Expected {n_features_out} feature names, got {len(feature_names_out_default)}"
+
+
+def check_param_validation(name, estimator_orig):
+    # Check that an informative error is raised when the value of a constructor
+    # parameter does not have an appropriate type or value.
+    rng = np.random.RandomState(0)
+    X = rng.uniform(size=(20, 5))
+    X = _pairwise_estimator_convert_X(X, estimator_orig)
+    y = rng.randint(0, 2, size=20)
+    y = _enforce_estimator_tags_y(estimator_orig, y)
+
+    # this object does not have a valid type for sure for all params
+    param_with_bad_type = type("BadType", (), {})()
+
+    fit_methods = ["fit", "partial_fit", "fit_transform", "fit_predict"]
+    methods = [method for method in fit_methods if hasattr(estimator_orig, method)]
+
+    for param_name in estimator_orig.get_params(deep=False):
+        match = rf"{param_name} must be .* Got .* instead."
+        err_msg = (
+            f"{name} does not raise an informative error message when the "
+            f"parameter {param_name} does not have a valid type or value."
+        )
+
+        estimator = clone(estimator_orig)
+
+        # First, check that the error is raised if param doesn't match any valid type.
+        estimator.set_params(**{param_name: param_with_bad_type})
+
+        for method in methods:
+            with raises(ValueError, match=match, err_msg=err_msg):
+                getattr(estimator, method)(X, y)
+
+        # Then, for constraints that are more than a type constraint, check that the
+        # error is raised if param does match a valid type but does not match any valid
+        # value for this type.
+        constraints = estimator_orig._parameter_constraints[param_name]
+        constraints = [make_constraint(constraint) for constraint in constraints]
+
+        for constraint in constraints:
+            if not hasattr(constraint, "generate_invalid_param_val"):
+                continue
+
+            bad_value = constraint.generate_invalid_param_val()
```

Comment:
I find having a public `generate_invalid_param_val` as a test fixutre be defined on `_Constraints` a bit counter-intuitive especially given that not all `_Constraints` define this method.

Apart from this test, is it used in other places? Would it make sense to make it an utility function?

Reply:
Transformed it into a function